crates/key-value/src/lib.rs

```diff
@@ -0,0 +1,183 @@
+}
+
+#[async_trait]
+trait Impl: Sync + Send {
+    async fn open(&mut self, name: &str) -> Result<Box<dyn ImplStore>, Error>;
+}
+
+#[async_trait]
+trait ImplStore: Sync + Send {
+    async fn get(&self, key: &str) -> Result<Vec<u8>, Error>;
+
+    async fn set(&self, key: &str, value: &[u8]) -> Result<(), Error>;
+
+    async fn delete(&self, key: &str) -> Result<(), Error>;
+
+    async fn exists(&self, key: &str) -> Result<bool, Error>;
+}
+
+pub struct KeyValueDispatch {
+    impls: HashMap<String, Box<dyn Impl>>,
+    stores: Table<Box<dyn ImplStore>>,
+}
+
+impl KeyValueDispatch {
+    fn new(config: Config) -> Self {
+        Self {
+            impls: config
+                .configs
+                .into_iter()
+                .map(|(name, config)| {
+                    (
+                        name,
+                        match config {
+                            ImplConfig::Sqlite(config) => {
+                                Box::new(sqlite::KeyValueSqlite::new(config)) as Box<dyn Impl>
+                            }
+                        },
+                    )
+                })
+                .collect(),
+            stores: Table::new(),
+        }
+    }
+}
+
+#[async_trait]
+impl KeyValue for KeyValueDispatch {
+    async fn open(&mut self, name: &str) -> Result<Store, Error> {
+        self.stores
+            .push(
+                self.impls
+                    .get_mut(name)
+                    .ok_or(Error::NoSuchStore)?
+                    .open(name)
+                    .await?,
+            )
+            .map_err(|()| Error::StoreTableFull)
+    }
+
+    async fn get(&mut self, store: Store, key: &str) -> Result<Vec<u8>, Error> {
+        self.stores
+            .get(store)
+            .ok_or(Error::InvalidStore)?
+            .get(key)
+            .await
+    }
+
+    async fn set(&mut self, store: Store, key: &str, value: &[u8]) -> Result<(), Error> {
+        self.stores
+            .get(store)
+            .ok_or(Error::InvalidStore)?
+            .set(key, value)
+            .await
+    }
+
+    async fn delete(&mut self, store: Store, key: &str) -> Result<(), Error> {
+        self.stores
+            .get(store)
+            .ok_or(Error::InvalidStore)?
+            .delete(key)
+            .await
+    }
+
+    async fn exists(&mut self, store: Store, key: &str) -> Result<bool, Error> {
+        self.stores
+            .get(store)
+            .ok_or(Error::InvalidStore)?
+            .exists(key)
+            .await
+    }
+
+    async fn close(&mut self, store: Store) {
+        self.stores.remove(store);
+    }
+}
+
+fn log_error(err: impl std::fmt::Debug) -> Error {
+    tracing::warn!("SQLite error: {err:?}");
+    Error::Runtime(format!("{err:?}"))
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[tokio::test(flavor = "multi_thread", worker_threads = 1)]
+    async fn all() -> Result<()> {
+        let mut kv = KeyValueDispatch::new(Config {
```

Comment:
is it worth using `tmpdir` to also test file path based Sqlite db rather than just in memory?

Reply:
I'd say no, since it's really SQLite's responsibility to ensure that in-memory databases function the same as disk-backed ones (minus durability, of course).  We don't need to test that for them -- they've already done it.